### PR TITLE
Mark Statement::execute() private, remove QueryBuilder::execute()

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes

See https://github.com/doctrine/dbal/pull/4580.

### CODE REVIEW

File: tests/Query/QueryBuilderTest.php [hunk 1]
```diff
@@ -849,23 +849,6 @@ public function testJoinWithNonUniqueAliasThrowsException(): void
...
-            ->expects(self::any())
-            ->method('executeQuery')
-            ->willReturn($this->createMock(Result::class));
-
-        $result = $qb
-            ->select('id')
-            ->from('foo')
-            ->execute();
-
-        self::assertInstanceOf(Result::class, $result);
```
Comment (morozov) [inline tests/Query/QueryBuilderTest.php:866]: This expectation is now enforced at the API level.

File: psalm.xml.dist [hunk 2]
```diff
@@ -55,15 +55,6 @@
...
                 <file name="tests/Functional/LegacyAPITest.php"/>
-                <!--
-                    These suppressions should be removed in 4.0.0
-                -->
-                <referencedMethod name="Doctrine\DBAL\Query\QueryBuilder::execute"/>
-                <referencedMethod name="Doctrine\DBAL\Statement::execute"/>
-                <!--
-                    TODO: remove in 4.0.0
-                -->
-                <referencedMethod name="Doctrine\DBAL\Schema\Schema::getMigrateToSql"/>
```
Comment (morozov) [inline psalm.xml.dist:66]: This one is a leftover from a previous deprecated API removal.

File: src/Statement.php [hunk 3]
```diff
@@ -154,19 +153,11 @@ public function bindParam($param, &$variable, int $type = ParameterType::STRING,
     /**
      * {@inheritDoc}
      *
-     * @deprecated Statement::execute() is deprecated, use Statement::executeQuery() or executeStatement() instead
-     *
      * @throws Exception
      */
-    public function execute(?array $params = null): Result
+    private function execute(array $params = []): Result
```
Comment (derrabus) [inline src/Statement.php]:
```suggestion
    private function execute(array $params): Result
```

This private method is never called without parameters. Let's keep things simple. 🙂

File: src/Statement.php [hunk 4]
```diff
@@ -154,19 +153,11 @@ public function bindParam($param, &$variable, int $type = ParameterType::STRING,
...
-        Deprecation::triggerIfCalledFromOutside(
-            'doctrine/dbal',
-            'https://github.com/doctrine/dbal/pull/4580',
-            'Statement::execute() is deprecated, use Statement::executeQuery() or Statement::executeStatement() instead'
-        );
-
-        if ($params !== null) {
+        if ($params !== []) {
             $this->params = $params;
         }
```
Comment (derrabus) [inline src/Statement.php:162]:
```suggestion
        $this->params = $params;
```

Assigning an empty array should be safe here, as it has always been.

File: src/Statement.php (hunk 3 above)
Comment (derrabus) [reply src/Statement.php]: Oh, you've changed that already. 🚀

File: src/Statement.php (hunk 4 above)
Comment (morozov) [reply src/Statement.php:162]: I believe the idea is that if you're calling `execute()` without parameters, it doesn't override the parameters bound via `bindParam()` or `bindValue()`. Not sure how sane this API is but I believe this is inherited from PDO.

Comment (derrabus) [reply src/Statement.php:162]: I see, thanks.

**Review by derrabus — APPROVED** [review]

**Pull request merged**